File: 0624/SeoyeonHong/주차 요금 계산.py
from collections import defaultdict
import math

def string_to_minutes(string):
    total_minutes = 0
    hour, minute = string.split(":")
    total_minutes = int(hour) * 60 + int(minute)
    return total_minutes

# 자동차별 주차 요금 계산 
def solution(fees, records):
    answer = []
    base_time = fees[0]
    base_fee = fees[1]
    overage_time = fees[2]
    overage_fee = fees[3]
    
    enter_times = defaultdict(list)
    total_time = defaultdict(int)
    toatal_fees = defaultdict(int)
    
    for record in records:
        time, car, enter = record.split(" ")
        if enter == 'IN': # 입차
            enter_times[car].append(string_to_minutes(time))
            
        else: # 출차
            enter_times[car].append(string_to_minutes(time))
            
            
    for car in enter_times.keys():
        times = enter_times[car]
        if len(times) % 2 == 1: # 출차 기록이 없다면
            times.append(string_to_minutes("23:59"))
        for i in range(0, len(times), 2):
            total_time[car] += times[i+1] - times[i]

        if total_time[car] <= base_time: # 기본 시간 이하일 경우
            toatal_fees[car] = base_fee
        else: # 기본 시간 초과 시
            toatal_fees[car] = base_fee + overage_fee * math.ceil((total_time[car] - base_time) / overage_time)

    sorted_items = toatal_fees.items()
    sorted_items = sorted(sorted_items)
    answer = []

    for _, value in sorted_items:
        answer.append(value)

    return answer

File: 0624/HyowonPark/주차요금계산.js
function solution(fees, records) {
    const carTotalTimes = new Map();
    const carIn = new Map();
    const BASE_DATE = new Date();
    for (const [time, car, inOut] of records.map(e => e.split(' '))) {
        if (inOut === 'IN') {
            carIn.set(car, time);
        } else {
            if (carIn.has(car)) {
                const IN = new Date(BASE_DATE);
                IN.setHours(...carIn.get(car).split(':'));
                const OUT = new Date(BASE_DATE);
                OUT.setHours(...time.split(':'));
                const minutes = (OUT.getTime() - IN.getTime()) / 60000;
                if (carTotalTimes.has(car)) {
                    carTotalTimes.set(car, carTotalTimes.get(car) + minutes);
                } else {
                    carTotalTimes.set(car, minutes);
                }
                
                carIn.delete(car); // 출차한 차량 삭제!
            }
        }
    }

    // 출차 안 찍힌 차들 23:59로 처리
    for (const [car, time] of carIn.entries()) {
        const IN =  new Date(BASE_DATE);
        IN.setHours(...time.split(':'));
        const OUT = new Date(BASE_DATE);
        OUT.setHours(23, 59);
        const minutes = (OUT.getTime() - IN.getTime()) / 60000;
        if (carTotalTimes.has(car)) {
            carTotalTimes.set(car, carTotalTimes.get(car) + minutes);
        } else {
            carTotalTimes.set(car, minutes);
        }
    }


    return [...carTotalTimes.keys()].sort().map(e => carTotalTimes.get(e) <= fees[0] ? fees[1] : fees[1] 
    + Math.ceil((carTotalTimes.get(e) - fees[0]) / fees[2]) * fees[3]
    );
}


const r = solution([180, 5000, 10, 600]	, ["05:34 5961 IN", "06:00 0000 IN", "06:34 0000 OUT", "07:59 5961 OUT", "07:59 0148 IN", "18:59 0000 IN", "19:09 0148 OUT", "22:59 5961 IN", "23:00 5961 OUT"]);
console.log(r);